forrest.go

```diff
@@ -27,31 +85,229 @@ itter indicates weather to use iterative splitting for all categorical features
 with more then 6 categories.
 */
 
-func GrowRandomForest(fm *FeatureMatrix, target Target, candidates []int, nSamples int, mTry int, nTrees int, leafSize int, maxDepth int, splitmissing bool, force bool, vet bool, evaloob bool, importance *[]*RunningMean) *Forest {
-	f := &Forest{target.GetName(), make([]*Tree, 0, nTrees), 0.0}
+func GrowRandomForest(fm *FeatureMatrix, target Target, config *ForestConfig) *ForestModel {
+	var (
+		targetName = target.GetName()
+		allocs     = NewBestSplitAllocs(config.NSamples, target)
+		importance = NewRunningMeans(len(fm.Data))
+		nCases     = fm.Data[0].Length()
+		trees      = make([]*Tree, config.NTrees)
+		inbag      = make([][]float64, config.NSamples)
+		candidates = make([]int, 0, len(fm.Data)-1)
+	)
 
-	switch target.(type) {
-	case TargetWithIntercept:
-		f.Intercept = target.(TargetWithIntercept).Intercept()
+	// construct list of candidate features
+	for i, feature := range fm.Data {
+		if feature.GetName() != targetName {
+			candidates = append(candidates, i)
+		}
+	}
+
+	// instantiate inbag matrix
+	if config.InBag {
+		for i := range inbag {
+			inbag[i] = make([]float64, config.NTrees)
+		}
 	}
 
-	//Slices for reuse during search for best splitter.
-	allocs := NewBestSplitAllocs(nSamples, target)
+	for i := 0; i < config.NTrees; i++ {
+		var cases []int
+		if config.Replace {
+			cases = SampleWithReplacment(config.NSamples, nCases)
+		} else {
+			cases = SampleWithoutReplacement(config.NSamples, nCases)
+		}
 
-	for i := 0; i < nTrees; i++ {
-		nCases := fm.Data[0].Length()
-		cases := SampleWithReplacment(nSamples, nCases)
+		// if in-bag, keep track of the cases
+		if config.InBag {
+			for _, c := range cases {
+				inbag[c][i]++
+			}
+		}
 
 		tree := NewTree()
-		tree.Target = target.GetName()
-		f.Trees = append(f.Trees, tree)
-
-		f.Trees[i].Grow(fm, target, cases, candidates, nil, mTry, leafSize, maxDepth, splitmissing, force, vet, evaloob, false, importance, nil, allocs)
-		switch target.(type) {
-		case BoostingTarget:
-			ls, ps := f.Trees[i].Partition(fm)
-			f.Trees[i].Weight = target.(BoostingTarget).Boost(ls, ps)
-		}
+		tree.Target = targetName
+		trees[i] = tree
+		trees[i].Grow(
```

Comment:
[nit] maybe since `Grow` takes so many arguments from your `config`, maybe just pass along the `config`?